### ISSUE TITLE
CLIPBOARD:  needs testing on non-mac platforms

### ISSUE THREAD

**Issue by rmkaplan** [Interlisp/medley#211]
I pushed a new version that uses the mac pbpaste and pbcopy only if (UNIX-GETENV "OSTYPE") includes "darwin", otherwise defaults to "xclip" for both streams.

I don't know if this is correct, please test on all of your favorite platforms.

**Comment by masinter** [comment]
on Linux/wsl there is no "xclip", and I don't understand how it would work.
You need the OS Type of the X-server, not the medley process.

**Comment by rmkaplan** [comment]
Can you copy-paste between linux applications?

If not, then I would just change clipboard so that at least it works between different windows in the same medley, for internal consistency.

Otherwise, we have to figure out what the tie-in should be.

**Comment by nbriggs** [comment]
> ... I don't understand how it would work.
You need the OS Type of the X-server, not the medley process.

That is not correct.  The command is going to be executed on the system where the `lde` process is running.  The OS type of the server where the display is being presented is irrelevant.  The clip/copy/paste command communicates with the (potentially remote) X server to deal with the text.

**Comment by nbriggs** [comment]
On most Linux distributions you can install the `xclip` package.  Check and see if that's available under WSL.

**Comment by nbriggs** [comment]
If WSL is running Ubuntu, try "sudo apt install xclip"

**Comment by nbriggs** [comment]
In `CLIPBOARD.PASTE.STREAM` the call to `xclip` needs to be `xclip -o` (the default is `xclip -i`)

**Comment by nbriggs** [comment]
In fact, the best results may come from using the two commands
```
xclip -i -selection clipboard
xclip -o -selection clipboard
```
because then, if you are running on a remote machine, with display on a Mac, with clipboard/pasteboard sync turned on, the copy selections you make will be available in the Mac paste buffer to paste into any window not just within Medley.

**Comment by rmkaplan** [comment]
So it is still CREATE-PROCESS-STREAM, just with those elaborated arguments?

Is there a way of testing whether xclip has been enabled?
If it isn’t enabled, then I think the clipboard should just simulate an internal Medley stream.  And presumably not bother with Unicode conversion.

I’m unclear about the remote-machine scenarios.
	If I’m on my Mac (my display) but Medley is running on a remote machine:
		If I copy a medley string to the clipboard, would it be available to
			paste in other apps on the remote machine, or
			paste in other apps on my Mac
		If I am pasting into Medley, should I get strings only that were
			copied by another app on the remote machine, or
			copied by another app on my Mac



Also:  is it a reasonable assumption that all other external apps would traffic in utf-8/unicode strings, or should the format of the clipboard be a parameter of this little interface?

**Comment by nbriggs** [comment]
Yes, it's still a CREATE-PROCESS-STREAM, but instead of just plain "xclip", use the elaborated arguments.  The "-i ..." case is when you're sending TO xclip (the "pbcopy" case),  the "-o ..." when you're receiving from xclip (the "pbpaste" case). 

There isn't a low-cost principled way of deciding whether xclip is available.  I think perhaps that attempting to (CREATE-PROCESS-STREAM "nonexistant-executable") should return NIL rather than a useless stream object.

I’m unclear about the remote-machine scenarios.
	If I’m on my Mac (my display) but Medley is running on a remote machine:
		If I copy a medley string to the clipboard, would it be available to
			paste in other apps on the remote machine, or
			paste in other apps on my Mac
        -- it's available in the CMD-V paste buffer, so you can paste it into terminals connected to anywhere and to apps that are running with windows on your Mac.

		If I am pasting into Medley, should I get strings only that were
			copied by another app on the remote machine, or
			copied by another app on my Mac
        -- it will be anything that is in your paste buffer, so if you did a CMD-C in any window displaying on your Mac (that has selectable text in it) you will be able to META-V paste it into Medley.

The options in the XQuartz preferences say whether the X clipboard should be copied in and out of the macOS pasteboard (which is what pbcopy and pbpaste work on).  pbcopy and pbpaste have arguments that can change their behavior too, documented in the man pages.

I would assume that all other applications are trafficking in UTF-8 strings.

**Comment by masinter** [comment]
the days of non-UTF8 are over (except for legacy cases like ours).  The Unicode problem I'm worried about is the possibility of Lisp doing normalization when reading in character mode, but assigning "looks" by byte and not number of characters..  Also marking and marking tedit and lisp files with their charset 
I went to try out the clipboard 
[hello.txt](https://github.com/Interlisp/medley/files/6033490/hello.txt)
but where is the version you want others to test

**Comment by waywardmonkeys** [comment]
In the long run, I think that this needs to be a subr so that maiko can provide the correct implementation per platform. Running natively on macOS or Windows would need different code from running within X, as would running via WebAssembly in the browser.

**Comment by nbriggs** [comment]
Running natively on macOS should still be fine with `pbpaste` and `pbcopy`, shouldn't it?  Those know nothing about X.   We're just executing them to stuff strings into or retrieve strings from the native system pasteboard.  When lde isn't executing on a mac, and we're not running under X then we have a problem if we're trying to share strings with applications outside lde and there isn't a command line method to access the system pasteboard.  If Medley were just trying to implement copy/paste within the Lisp world it wouldn't bother with external programs.

**Comment by rmkaplan** [comment]
Test CLIPBOARD.LCOM (but I just pushed a new version)

The XCCS character sets that UNICODE loads by default are

(LATIN SYMBOLS1 SYMBOLS2 EXTENDED-LATIN FORMS SYMBOLS3 SYMBOLS4 ACCENTED-LATIN)

So test characters in some of those sets.

**Comment by rmkaplan** [comment]
Yes, the whole point is to let stanard copy/paste move strings to/from medley and other applications, in a lightweight way—without going through files.  What is getting a little murky is understanding the geographic relationship between Medley and which other applications are in scope.  Moving strings around inside Medley is a nice side-effect, for consistency, but it isn’t the motivation.  (Inside Medley it does have a nice little extra feature:  you can copy once and then paste in many locations, something you can’t do with the Medley copy-select convention.)

**Comment by masinter** [comment]
I tested on Windows/WSL and it seemed to work to copy things from outside medley to inside.
No luck the other way (copy from inside Medley to another windows or wsl/X app)

**Comment by rmkaplan** [comment]
OK, thanks.  Half way.

There may be something else to do to arm it the other direction, although I think I followed Nick’s instructions.

He may see the problem when he gets back.

**Comment by rmkaplan** [comment]
In terms of deciding whether/how the clipboard is reference (xclip, pbpaste), the suggestion was made to invoke uname, but then not to do that every time (although this would only be invoked when meta-C/V are pushed, so in the noise).  

A simple alternative would be for the run-medley to set a variable to the value of uname, and then clipboard could look at that variable.

**Comment by masinter** [comment]
uname has lots of options

       -s, --kernel-name
              print the kernel name

       -n, --nodename
              print the network node hostname

       -r, --kernel-release
              print the kernel release

       -v, --kernel-version
              print the kernel version

       -m, --machine
              print the machine hardware name

       -p, --processor
              print the processor type (non-portable)

       -i, --hardware-platform
              print the hardware platform (non-portable)

       -o, --operating-system
              print the operating system

I think fixing UNIX-GETPARM would be more straightforward.

**Comment by nbriggs** [comment]
See the uname table at Interlisp/maiko#129

**Comment by masinter** [comment]
after all of this discussion, copy/paste to/from Medley on a mac gives an error not finding xclip etc.   becaus GETENV("OSTYPE") is NIL with the latest medley startup scripts.

**Comment by rmkaplan** [comment]
Works for me, on a mac.  My .zshrc has a line at the top

export OSTYPE=$OSTYPE

Maybe that’s necessary?

**Comment by masinter** [comment]
@rmkaplan you may have missed https://github.com/Interlisp/maiko/pull/495

> Calls to UNIX-GETPARM with argument "OSNAME" or "ARCH" now return values derived from the compile-time settings in inc/maiko/platform.h. Relates to [Interlisp/medley#1596](https://github.com/Interlisp/medley/issues/1596)
> 
> The current return values are OSNAME:
> 
> * "macOS"
> * "Cygwin"
> * "DragonFly BSD"
> * "FreeBSD"
> * "Linux"
> * "NetBSD"
> * "OpenBSD"
> * "AmigaOS 3"
> * "Solaris"
> * "Windows"
> * "Emscripten"
> 
> ARCH:
> 
> * "WebAssembly"
> * "x86_64"
> * "arm"
> * "arm64"
> * "x86"
> * "PowerPC"
> * "RISC-V"
> * "SPARC"
> * "Motorola68K"

**Comment by rmkaplan** [comment]
Are you saying that CLIPBOARD now should be calling  (UNIX-GETPARM "OSNAME") instead of (UNIX-GETENV "OSTYPE") ?

Still doesn't explain why it works for me the way it is.

**Comment by nbriggs** [comment]
If the test is purely "are we running on macOS, or not" then (UNIX-GETPARM "OSNAME") would be the right thing.  If it's more complicated than that then we need to fill out the matrix of possibilities for what you need to call in which circumstance.

**Comment by masinter** [comment]
It's not clear what the "issue" is from this discussion.
There are some limitations of CLIPBOARD in some OSes, or some things need to be documented. But I can't see a solvable problem here without more analysis, so I'm converting this into a discussion
